File: Perceptron.java
public class Perceptron {

    private double[][] neuroInputs;
    private double[] weights;
    private double learningRate;
    private int[] expectedOutputs;
    private int epochs;

    public Perceptron() {
        this.neuroInputs = new double[][] {
            {0, 0},
            {0, 1},
            {1, 1},
            {1, 0}
        };
        this.weights = new double[] {0.0, 0.0};
        this.learningRate = 0.1;
        this.expectedOutputs = new int[] {0, 0, 1, 0};
        this.epochs = 0;
    }

    private int step(double valueActivation) {
        return valueActivation >= 1.0 ? 1 : 0;
    }

    private double sum(double[] neuroInput, double[] weights) {
        return neuroInput[0] * weights[0] + neuroInput[1] * weights[1];
    }

    public Result run() {
        while (true) {
            double totalError = 0;
            for (int i = 0; i < neuroInputs.length; i++) {
                double[] neuroInput = neuroInputs[i];
                double sumNeuro = sum(neuroInput, weights);
                int activation = step(sumNeuro);
                double error = Math.abs(activation - expectedOutputs[i]);
                totalError += error;
                for (int j = 0; j < weights.length; j++) {
                    weights[j] += learningRate * neuroInput[j] * error;
                }
            }
            epochs++;
            System.out.println(epochs);
            if (totalError == 0) {
                return new Result(weights, epochs);
            }
        }
    }

    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        Perceptron perceptron = new Perceptron();
        Result result = perceptron.run();
        long endTime = System.currentTimeMillis();
        double executionTime = (endTime - startTime) / 1000.0;
        System.out.printf("Execution Time: %.4f seconds%n", executionTime);
        System.out.printf("Weights: %.2f - %.2f%nEpochs: %d%n", result.weights[0], result.weights[1], result.epochs);
    }

    public static class Result {
        public final double[] weights;
        public final int epochs;

        public Result(double[] weights, int epochs) {
            this.weights = weights;
            this.epochs = epochs;
        }
    }
}

File: Perceptron.rb
require 'time'

class Perceptron
  def initialize
    @neuro_inputs = [
      [0, 0],
      [0, 1],
      [1, 1],
      [1, 0]
    ]
    @weights = [0.0, 0.0]
    @learning_rate = 0.1
    @expected_outputs = [0, 0, 1, 0]
    @epochs = 0
  end

  def step(value_activation)
    value_activation >= 1 ? 1 : 0
  end

  def sum(neuro_input, weights)
    neuro_input[0] * weights[0] + neuro_input[1] * weights[1]
  end

  def run
    loop do
      total_error = 0
      @neuro_inputs.each_with_index do |neuro_input, index|
        sum_neuro = sum(neuro_input, @weights)
        activation = step(sum_neuro)
        error = (activation - @expected_outputs[index]).abs
        total_error += error
        @weights.each_with_index do |weight, index_|
          @weights[index_] += @learning_rate * neuro_input[index_] * error
        end
      end
      @epochs += 1
      puts @epochs
      break if total_error.zero?
    end
    { weights: @weights, epochs: @epochs }
  end
end

start_time = Time.now
perceptron = Perceptron.new
answer = perceptron.run
end_time = Time.now
execution_time = end_time - start_time
puts format('Execution Time: %.4f seconds', execution_time)
puts format('Weights: %.2f - %.2f\nEpochs: %d', answer[:weights][0], answer[:weights][1], answer[:epochs])
